### ISSUE TITLE
[1.3-beta.16] meteorEnv is undefined

### ISSUE THREAD

**Issue by ne8il** [meteor/meteor#6459]
(Moved from comment on other thread to own issue)

I've upgraded from 1.3-beta.12 (which was working) to 1.3-beta.16. I get a client-side error when I load the app in a browser because `meteorEnv` is `undefined`. When I look at where that gets assigned, I see that `Package["meteor-env-dev"]` = `Object {meteorEnv: undefined}`. Am I supposed to be setting that somewhere?

To clarify, the line that breaks is `client_environment.js`, line 12:  
```javascript
isProduction: meteorEnv.NODE_ENV === "production"
```

Where I inspected to see where that was assigned is `import_meteorEnv.js`, line 5-8.

Here's the value of `__meteor_runtime_config__`: 
```javascript
{
"meteorRelease":"METEOR@1.3-beta.16",
"PUBLIC_SETTINGS":{...},
"ROOT_URL":"http://localhost:3000/",
"ROOT_URL_PATH_PREFIX":"",
"accountsConfigCalled":true,
"appId":"...",
"autoupdateVersion":"b043cd314df097368760e28dea3d05f7ee67bb02",
"autoupdateVersionRefreshable":"7b78d68809b23ee39eb727fefb083a8b3f027938",
"autoupdateVersionCordova":"none"
}
```

**Comment by benjamn** [comment]
I haven't been able to reproduce this by creating an empty app with `beta.12` and upgrading it to `beta.16`. Can you provide some reproduction steps? Alternatively, does `meteor reset` fix the problem?

**Comment by tmeasday** [comment]
@ne8il how are you running your app? (The exact command line)

**Comment by ne8il** [comment]
@tmeasday Sure

```
meteor run --settings settings.dev.json
```

This does start velocity but I also ran `VELOCITY=0 meteor run --settings settings.dev.json` and got the same results.

**Comment by tmeasday** [comment]
Hmm. This seems quite surprising, given that code. We are re-working this code at the moment however, so it may not be worth investigating further until the next beta is out.

OTOH if you can post a simple reproduction somewhere I'd love to look at it.

**Comment by Tarang** [comment]
I upgraded from beta.11 to beta.16 and got the same problem. It gets stuck on the splash screen and if I inspect the code/refresh the page it boils down to the same meteorEnv error.

This doesn't happen loading up http://localhost:3000 in a web browser, only on the cordova build. I don't use a meteor settings bundle.

![screen shot 2016-03-12 at 10 48 35 am](https://cloud.githubusercontent.com/assets/1616628/13722010/0c83d760-e840-11e5-9a92-b7e3553e9605.png)

**Comment by Tarang** [comment]
Reproduction:

```
meteor create project
cd project
meteor update --release 1.3-beta.16
meteor add-platform ios
meteor run ios
```

**Comment by benjamn** [comment]
Thanks for the reproduction, @Tarang. I'm confident this will be fixed in the release candidate (coming soon), due to this commit: https://github.com/meteor/meteor/commit/d87dddfc9b79b01791fadf405d18358ef59a262a

**Comment by Tarang** [comment]
@benjamn Will this make the next beta?

**Comment by insaneinc01** [comment]
`App startup timed out, reverting to last known good version`
Waiting for the next upgrade. :|

**Comment by denisbabineau** [comment]
@benjamn I'm not sure this is fixed by d87dddf I'm running from git ( currently on 322e86c ) and I'm still running into this problem on iOS. `__meteor_runtime_config__.meteorEnv` is set properly in the browser environment but not in cordova. From what I can tell the boilerplate for browser is generated at runtime but for cordova it's at build time (probably makes sense, to store the template in the bundle?). Anyhow, I added log statements here:

packages/webapp/webapp_server.js@294:
```
  console.log(`!!!! meteorRuntimeConfig in webapp_server.js: ${JSON.stringify(runtimeConfig)}`);
```

tools/cordova/builder.js@419:
```
    console.log(`!!!! meteorRuntimeConfig in cordova/builder.js: ${JSON.stringify(runtimeConfig)}`);
```

And got the following output when running `gitmeteor run ios-device --port 3500`:

during the build process (notice the missing meteorEnv):
```
!!!! meteorRuntimeConfig in cordova/builder.js: {"meteorRelease":"none","ROOT_URL":"http://192.168.2.50:3500/","ROOT_URL_PATH_PREFIX":"","DDP_DEFAULT_CONNECTION_URL":"http://192.168.2.50:3500/","autoupdateVersionCordova":"c915c78ec4276da188be5dda33e4189df9c0f8e5","appId":"elwi3g1dfxkfd2oc70b"}
```

and once the server started (with meteorEnv):
```
=> Started app on iOS Device.
I20160314-12:56:07.393(-3)? !!!! meteorRuntimeConfig in webapp_server.js: {"meteorRelease":"none","PUBLIC_SETTINGS":{},"ROOT_URL":"http://localhost:3500/","ROOT_URL_PATH_PREFIX":"","appId":"elwi3g1dfxkfd2oc70b","autoupdateVersion":"53b68fb467ffb6a057ccc4a6fffc46e08a91ff8b","autoupdateVersionRefreshable":"f3a567ae951c556e50bc0c518d8dcba1ae7e9a94","autoupdateVersionCordova":"c915c78ec4276da188be5dda33e4189df9c0f8e5","meteorEnv":{"NODE_ENV":"development","TEST_METADATA":"{}"}}
=> Started your app.                          
                                              
=> App running at: http://localhost:3500/     
I20160314-12:56:07.446(-3)? !!!! meteorRuntimeConfig in webapp_server.js: {"meteorRelease":"none","PUBLIC_SETTINGS":{},"ROOT_URL":"http://192.168.2.50:3500/","ROOT_URL_PATH_PREFIX":"","appId":"elwi3g1dfxkfd2oc70b","autoupdateVersion":"53b68fb467ffb6a057ccc4a6fffc46e08a91ff8b","autoupdateVersionRefreshable":"f3a567ae951c556e50bc0c518d8dcba1ae7e9a94","autoupdateVersionCordova":"c915c78ec4276da188be5dda33e4189df9c0f8e5","DDP_DEFAULT_CONNECTION_URL":"http://192.168.2.50:3500/","meteorEnv":{"NODE_ENV":"development","TEST_METADATA":"{}"}}
```

**Comment by benjamn** [comment]
@martijnwalraven are you familiar with code that deals with `__meteor_runtime_config__` in Cordova?

**Comment by martijnwalraven** [comment]
Yes, there are two places where `__meteor_runtime_config__` is set for Cordova. One is during the initial build (https://github.com/meteor/meteor/blob/release-1.3/tools/cordova/builder.js#L405-L413) and the other is when serving a new bundle for a hot code push (https://github.com/meteor/meteor/blob/release-1.3/packages/webapp/webapp_server.js#L542-L574).

**Comment by benjamn** [comment]
Fixed in `1.3-rc.1`, thanks to https://github.com/meteor/meteor/commit/9411f0a80e584f0a40612b920449a96e103b6514.

**Comment by ne8il** [comment]
:+1: Thanks, will give it a shot this afternoon.

**Comment by insaneinc01** [comment]
Weirdly, this does not work for me even on `METEOR@1.3-rc.1`.
The problem is only when I push it to AppStore via TestFlight for Adhoc distribution. Its stuck on the splash screen. 
The app installs on the phone without problems from xcode directly in dev mode. I guess this part is fixed with 1.3-rc.1

Is anyone else facing the same problem?

**Comment by Tarang** [comment]
@insaneinc01 In your testing environment make sure you've deleted the old app before installing the new one.

**Comment by insaneinc01** [comment]
@Tarang No luck, ran `meteor reset` as well as made sure there were no remnants of the previous version.. still the same issue. 
Updating to rc.2 now. Lets see how this goes.

**Comment by martijnwalraven** [comment]
`rc.2` has no real changes apart from Windows support, so I'm afraid that probably won't make a difference.

Could you try running from Xcode with `meteor run ios-device --production` and see if you can reproduce the problem that way?

**Comment by Tarang** [comment]
@insaneinc01 Not `meteor reset`. If you don't delete the app on the environment having the issue first it will use the previous apps bundle containing the issue and you'll get that freeze at the splash screen.

**Comment by insaneinc01** [comment]
Hello @martijnwalraven 

I tried `meteor run ios-device --production`. Works well when I install the app directly from Xcode. 
The problem is only on submission to AppStore. The 'resulting API analysis file is too large' warning always shows up. After the usual appstore processing, when I install it on the device via TestFlight, the app hangs on splash screen. After I kill it and re-launch, it never opens and immediately crashes.   

I also tried with `meteor build ../app-build --server http://server-url --verbose`
Xcode throws this error - `process launch failed: timed out waiting for app to launch`

So it looks like there's a problem with only the **'built'** version of the app.
